Building OpenBLAS 0.3.24 fails on MacPorts running macOS 13.5.2

# Description

Updating to OpenBLAS 0.3.24 on MacPorts seg faults due to invalid memory reference. Issue [reported to MacPorts](https://trac.macports.org/ticket/68225#comment:3) with response from MacPorts maintainers to escalate upstream to OpenBLAS.

Error message below:

```text
OPENBLAS_NUM_THREADS=1 OMP_NUM_THREADS=1 ./cblat2 < ./cblat2.dat

Program received signal SIGSEGV: Segmentation fault - invalid memory reference.

Backtrace for this error:

Program received signal SIGTRAP: Trace/breakpoint trap.

Backtrace for this error:
OPENBLAS_NUM_THREADS=1 OMP_NUM_THREADS=1 ./zblat2 < ./zblat2.dat
```


# Environment

Apple M1 Pro, macOS 13.5.2 (22G91)


# Steps to Reproduce

Upgrade OpenBLAS from 0.3.23\_0 to 0.3.24\_0


# Expected Result

OpenBLAS 0.3.24 is built and installed.


# Actual Result

Build hangs, no backtrace is emitted in build log.

[openblas-upgrade.log](https://github.com/OpenMathLib/OpenBLAS/files/12718568/openblas-upgrade.log)

I am facing the same issue and was able to reproduce the issue just running `make` in a checked-out git master. It seems to be related to Apple Clang. There was very recently an XCode update (15.0) which could play a role here. By default, on my system the following compilers are used:

```
% cc --version
Apple clang version 15.0.0 (clang-1500.0.40.1)
Target: arm64-apple-darwin22.6.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin

% gfortran --version
GNU Fortran (MacPorts gcc12 12.3.0_0+stdlib_flag) 12.3.0
Copyright (C) 2022 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

When building with `CC=clang make`, I can force the use of macport's clang instead of the system one:
```
% clang --version
clang version 16.0.6
Target: arm64-apple-darwin22.6.0
Thread model: posix
InstalledDir: /opt/local/libexec/llvm-16/bin
```

With that, the segfault is gone. However, the tests fails at a later point in time:
```
TEST 38/38 kernel_regress:skx_avx dyld[9121]: missing symbol called
make[1]: *** [run_test] Abort trap: 6
make: *** [tests] Error 2
```

Update on more versions:

* `CC=/opt/local/bin/clang-mp-17 make NO_AVX=1`: does not compile (see below)
* `CC=/opt/local/bin/clang-mp-16 make NO_AVX=1`: only failing in the skx_avx test
* `CC=/opt/local/bin/clang-mp-15 make NO_AVX=1`: **segfault**
* `CC=/opt/local/bin/clang-mp-14 make NO_AVX=1`: **segfault**
* `CC=/opt/local/bin/clang-mp-13 make NO_AVX=1`: **segfault**
* `CC=/opt/local/bin/clang-mp-12 make NO_AVX=1`: only failing in the skx_avx test

So clang versions 13, 14 and 15 show the error for me. I am a bit surprised that this issue only now popped up as I would think that XCode 14 came with a newer clang than version 12. And it also contradicts the CI that successfully builds with clang 14.

With clang 17, I get errors during compilation:
```
In file included from ../include/lapacke_config.h:43:
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/stdlib.h:66:10: fatal error: 'sys/wait.h' file not found
   66 | #include <sys/wait.h>
      |          ^~~~~~~~~~~~
In file included from lapacke_cheev_2stage.c:33:
In file included from ../include/lapacke_utils.h:36:
In file included from ../include/lapacke.h:36:
In file included from ../include/lapack.h:8:
In file included from ../include/lapacke_config.h:43:
In file included from /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/stdlib.h:128:
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/malloc/_malloc.h:38:10: fatal error: 'malloc/_malloc_type.h' file not found
   38 | #include <malloc/_malloc_type.h>
      |          ^~~~~~~~~~~~~~~~~~~~~~~
```

Those files are present though, so I think this may be an issue with clang 17 from macports.

--- a/Makefile.system
+++ b/Makefile.system
@@ -405,6 +405,13 @@ export MACOSX_DEPLOYMENT_TARGET=10.8
 endif
 endif
 MD5SUM = md5 -r
+XCVER = $(shell pkgutil --pkg-info=com.apple.pkg.Xcode |awk '/version:/ {print $2}'|cut -d: -f2|cut -f1 -d.)
+ifeq (x$(XCVER)x,xx)
+XCVER = $(shell pkgutil --pkg-info=com.apple.pkg.CLTools_Executables |awk '/version:/ {print $2}'|cut -d: -f2|cut -f1 -d.)
+endif
+ifeq (x$(XCVER), x 15)
+CCOMMON_OPT += -Wl,-ld_classic
+endif
 endif
 
 ifneq (,$(findstring $(OSNAME), FreeBSD OpenBSD DragonFly))

